### PR TITLE
Feature - Update and draw map

### PR DESCRIPTION
**Pull request by at-hoavo**
#### Related issues (Optional)
Task #54

#### Description (Required)
- [x] Update map when click item on RecyclerView
- [x] Combine arrived screen into map

### CODE REVIEW

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+import vn.asiantech.way.ui.arrived.DialogArrived
+import vn.asiantech.way.ui.base.BaseActivity
+import vn.asiantech.way.ui.custom.FloatingButtonHorizontal
+import vn.asiantech.way.ui.home.HomeAdapter
+import vn.asiantech.way.ui.register.RegisterActivity
+import vn.asiantech.way.ui.search.SearchLocationActivity
+import vn.asiantech.way.ui.share.ShareLocationActivity
+import vn.asiantech.way.utils.LocationUtil
+
+/**
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: One class should have one responsibility. Large classes tend to handle many things at once. Split up large classes into smaller classes that are easier to understand.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Classes with many functions tend to do too many things and often come in conjunction with large classes and can quickly become God classes. Consider extracting methods to (new) classes better matching their responsibility.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+        frOverlay.visibility = View.GONE
+    }
+
+    private fun drawCurrentMaker(location: android.location.Location) {
+        if (mGoogleMap != null) {
+            mGoogleMap?.clear()
+            val currentLocation = LatLng(location.latitude, location.longitude)
+            addMarker(MarkerOptions()
+                    .position(currentLocation)
+                    .draggable(true)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        polyPointOption.width(TYPE_POLYLINE_WIDTH)
+        mGoogleMap?.addPolyline(polyPointOption)
+    }
+
+    private fun updateMap(points: List<LatLng>) {
+        mGoogleMap?.clear()
+        addMarker(setMarkerOption(R.drawable.ic_ht_source_place_marker, mBegin), mBegin)
+        if (points.size > 1) {
+            if (points[points.size - 1] != mDestination) {
+                addMarker(setMarkerOption(R.drawable.ic_rectangle, points[points.size - 1]), points[points.size - 1])
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+            mLocations[it].isChoose = true
+            mHomeAdapter.notifyItemChanged(it)
+            mPosition = it
+            updateMap(setListToPosition(it))
+        }
+        recycleViewLocation.layoutManager = LinearLayoutManager(this)
+        recycleViewLocation.adapter = mHomeAdapter
+    }
+
+    private fun setListToPosition(position: Int): List<LatLng> = (0..position).map { mLocations[it].point }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    private fun initMap() {
+        val supportMapFragment = supportFragmentManager.findFragmentById(R.id.fragmentMap) as? SupportMapFragment
+        supportMapFragment?.getMapAsync(this)
+    }
+
+    private fun setPaddingGoogleLogo() {
+        val display = windowManager.defaultDisplay
+        val size = Point()
+        display.getSize(size)
+        mGoogleMap?.setPadding(PADDING_LEFT, PADDING_TOP, PADDING_RIGHT, size.y / 3)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+//        mLocations.add(Location("4:00 PM", "Destination", "1 hour ago | 5km", LatLng(16.0725051, 108.2296716)))
+//        mLocations.add(Location("5:00 PM", "Stop", "30 minutes | 1km", LatLng(16.0717437, 108.2236926)))
+//        mLocations.add(Location("6:00 PM", "Start", "15 minutes| 5km", LatLng(16.0712047, 108.2193197)))
+        mArrived.latLngs = mutableListOf()
+        mLocations.forEach {
+            it.point.let {
+                mArrived.latLngs?.add(it)
+            }
+        }
+        mBegin = LatLng(16.0721115, 108.2302225)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+        mDestination = LatLng(16.0712047, 108.2193197)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+    override fun onBackPressed() {
+        if (isExit) {
+            val intent = Intent(Intent.ACTION_MAIN)
+            intent.addCategory(Intent.CATEGORY_HOME)
+            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
+            startActivity(intent)
+        } else {
+            toast(resources.getString(R.string.register_double_click_to_exit))
+            isExit = true
+            Handler().postDelayed({ isExit = false }, 3 * 1000)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt [hunk 9]
```diff
@@ -160,13 +160,12 @@ class HomeActivity : BaseActivity(), OnMapReadyCallback, FloatingButtonHorizonta
...
 
     private fun initDummyData(locations: ArrayList<Location>) {
-        locations.add(Location("1:00 PM", "Stop", "30 minutes| You stop at the school..............."))
-        locations.add(Location("2:00 PM", "Drive", "50 minutes| You went to the Hoa Khanh market........"))
-        locations.add(Location("3:00 PM", "Walk", "30 minutes | 1km"))
-        locations.add(Location("4:00 PM", "Destination", "1 hour ago | 5km"))
-        locations.add(Location("5:00 PM", "Stop", "30 minutes | 1km"))
-        locations.add(Location("6:00 PM", "Start", "15 minutes| 5km"))
-        locations.add(Location("7:00 PM", "Moto", "40 minutes | 3km"))
+        locations.add(Location("1:00 PM", "Stop", "30 minutes| You stop at the school...............", LatLng(16.0721115, 108.2302225)))
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt [hunk 10]
```diff
@@ -160,13 +160,12 @@ class HomeActivity : BaseActivity(), OnMapReadyCallback, FloatingButtonHorizonta
...
+        locations.add(Location("2:00 PM", "Drive", "50 minutes| You went to the Hoa Khanh market........", LatLng(16.0712047, 108.2193197)))
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt [hunk 11]
```diff
@@ -160,13 +160,12 @@ class HomeActivity : BaseActivity(), OnMapReadyCallback, FloatingButtonHorizonta
...
+        locations.add(Location("3:00 PM", "Walk", "30 minutes | 1km", LatLng(16.0721611, 108.2303906)))
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt [hunk 12]
```diff
@@ -160,13 +160,12 @@ class HomeActivity : BaseActivity(), OnMapReadyCallback, FloatingButtonHorizonta
...
+        locations.add(Location("4:00 PM", "Destination", "1 hour ago | 5km", LatLng(16.0725051, 108.2296716)))
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt [hunk 13]
```diff
@@ -160,13 +160,12 @@ class HomeActivity : BaseActivity(), OnMapReadyCallback, FloatingButtonHorizonta
...
+        locations.add(Location("5:00 PM", "Stop", "30 minutes | 1km", LatLng(16.0717437, 108.2236926)))
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt [hunk 14]
```diff
@@ -160,13 +160,12 @@ class HomeActivity : BaseActivity(), OnMapReadyCallback, FloatingButtonHorizonta
...
+        locations.add(Location("6:00 PM", "Start", "15 minutes| 5km", LatLng(16.0712047, 108.2193197)))
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/home/HomeActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt (hunk 1 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: One class should have one responsibility. Large classes tend to handle many things at once. Split up large classes into smaller classes that are easier to understand.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Classes with many functions tend to do too many things and often come in conjunction with large classes and can quickly become God classes. Consider extracting methods to (new) classes better matching their responsibility.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+//        mLocations.add(Location("1:00 PM", "Stop", "30 minutes| You stop at the school...............", LatLng(16.0721115, 108.2302225)))
+//        mLocations.add(Location("3:00 PM", "Walk", "30 minutes | 1km", LatLng(16.0721611, 108.2303906)))
+//        mLocations.add(Location("4:00 PM", "Destination", "1 hour ago | 5km", LatLng(16.0725051, 108.2296716)))
+//        mLocations.add(Location("5:00 PM", "Stop", "30 minutes | 1km", LatLng(16.0717437, 108.2236926)))
+//        mLocations.add(Location("6:00 PM", "Start", "15 minutes| 5km", LatLng(16.0712047, 108.2193197)))
+        mArrived.latLngs = mutableListOf()
+        mLocations.forEach {
+            it.point.let {
+                mArrived.latLngs?.add(it)
+            }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+        }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+import vn.asiantech.way.ui.share.ShareLocationActivity
+import vn.asiantech.way.utils.LocationUtil
+
+/**
+ * Copyright © 2017 Asian Tech Co., Ltd.
+ * Created by at-hoavo on 25/10/2017.
+ */
+internal class UpdateMap : BaseActivity(), OnMapReadyCallback,
+        FloatingButtonHorizontal.OnMenuClickListener {
+    companion object {
```
Comment (at-quocnguyen) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: break line cho đep + dễ nhìn

**Comment by at-quocnguyen** [comment]
luôn luôn check nhánh của mình với master nhé !

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt (hunk 17 above)
Comment (at-hoavo) [reply app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: dạ cái này là em làm cái  follow khi click vào item trên recyclerView thì update lại vị trí trên map và vẽ lại cái đường mà nó đã đi, với cả khi mà đã đi tới điểm đích thì show thông tin đó a

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+import vn.asiantech.way.ui.arrived.DialogArrived
+import vn.asiantech.way.ui.base.BaseActivity
+import vn.asiantech.way.ui.custom.FloatingButtonHorizontal
+import vn.asiantech.way.ui.home.HomeAdapter
+import vn.asiantech.way.ui.register.RegisterActivity
+import vn.asiantech.way.ui.search.SearchLocationActivity
+import vn.asiantech.way.ui.share.ShareLocationActivity
+import vn.asiantech.way.utils.LocationUtil
+
+/**
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: One class should have one responsibility. Large classes tend to handle many things at once. Split up large classes into smaller classes that are easier to understand.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Classes with many functions tend to do too many things and often come in conjunction with large classes and can quickly become God classes. Consider extracting methods to (new) classes better matching their responsibility.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+            mLocations[it].isChoose = true
+            mHomeAdapter.notifyItemChanged(it)
+            mPosition = it
+            updateMap(setListToPosition(it))
+        }
+        recycleViewLocation.layoutManager = LinearLayoutManager(this)
+        recycleViewLocation.adapter = mHomeAdapter
+    }
+
+    private fun setListToPosition(position: Int): List<LatLng> =
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+            (0..position).map { mLocations[it].point }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

**Comment by at-hoavo** [comment]
@at-quocnguyen please check for me, thanks!

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt (hunk 20 above)
Comment (at-hoavo) [reply app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: cái này dòng của e ko dài mà ci vẫn bắt a

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt (hunk 18 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: One class should have one responsibility. Large classes tend to handle many things at once. Split up large classes into smaller classes that are easier to understand.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Classes with many functions tend to do too many things and often come in conjunction with large classes and can quickly become God classes. Consider extracting methods to (new) classes better matching their responsibility.

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt (hunk 19 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt (hunk 20 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/update/UpdateMap.kt]: WARNING: Line detected that is longer than the defined maximum line length in the code style.

**Pull request merged**